### ISSUE TITLE
Should we capture provenance for the control plane?

### ISSUE THREAD

**Issue by marcelamelara** [slsa-framework/slsa#1253]
Does this mean that there is no provenance for the control plane? Should there be (except for the bootstrapping problem)?

_Originally posted by @arewm in https://github.com/slsa-framework/slsa/pull/1244#discussion_r1869470320_

**Comment by marcelamelara** [comment]
CC @paveliak

**Comment by paveliak** [comment]
That's an interesting problem. The short answer is "no" and let me elaborate...

Ultimately the objective of Build Environment track is to provide cryptographic evidence that build is running using the specified image. Image itself has provenance and comes from a build pipeline with a SLSA level but since build runs _inside_ the environment (aka "inside the image") we need to rely on extra things to actually attest the environment build runs in.

One option to address that could be to trust control plane to do that and attest the control plane instead. Apple is doing some very interesting work in this area with their [Private Compute Cloud](https://security.apple.com/documentation/private-cloud-compute). Below I enumerate few challenges that existing build platforms would need to provide answer for and to be honest I think that the bar could be prohibitively high for the majority of them.

In this track we take slightly different approach relying on the hardware root of trust to do the attestation leaving control plane outside of `TCB`.  This allows us to scope attestation to a single machine instead of attesting Control plane (generally a cloud service) that could consist of multiple machines and micro-services. But the downside is we rely on hardware capabilities (mostly `SecureBoot` and `TPM` or similar hardware) which may not be present on all platforms.

Another caveat is by treating Control plane as untrusted we have to accept security risks coming from the interaction with the Control plane. For example, we have to trust build request message that initiates the build. Granted, the important properties of this message are included into the build attestation (e.g. commit SHA). And generally we should include into the provenance all input messages coming from the control plan that can influence build results. 

I would add another interesting consideration regarding untrusted control plane as we are on this topic. As we boot build environment we cannot expect (untrusted) control plane to attest build environment before executing the job and terminate environment if it failed the attestation. Rather we include hardware-assisted build environment attestation into the build provenance and let build consumer decide on what to do. 

Getting back to attesting Control plane...  Suppose we request attestation of the Control plane. What kind of challenges come with this request for the build platform providers?
1) There has to be a transparency with the Control plane code. Digests for all builds (and perhaps more than that) should be published in a transparency log. It is not uncommon to make changes to the Cloud services multiple times a day and we can expect constant churn in that log.
2) There has to be a way to maintain request affinity. Cloud services usually run on many nodes (e.g. containers on K8s cluster) and every request may land on a different node. If client attested node 1 and subsequent request lands on node 2 then prior attestation does not make much sense
3) There has to be some guarantees about the inability to access Control plane via a back door (for administrative purpose)
4) Control plane is a stateful service and attesting compute nodes won't be enough, we would need to attest storage nodes too (and technically all the transitive dependencies)
5) Persistent storage introduces the notion of time into the picture. Not only we would need to attest current state of Storage but also give some guarantees that yesterday (and before that) it was in good shape too.

I would be happy to see these questions solved as they may be relevant to other tracks too (Source track? tagging @zachariahcox for it).

**Comment by bureado** [comment]
My read of the `BuildEnv` levels is that L1 distrusts image stores, L2 distrusts the control plane, and L3 distrusts the host. The security effects vary. In L1, a correct build platform will refuse to build if an artifact store has tampered with the build image, but a user has to trust the platform is correct. In L2, the user can get proof from the OS that what's measured is what was expected (what was in the `BuildEnv` provenance), but the user has to trust the host is not controlling the hypervisor, memory, vTPM implementation. In L3, the user can get that proof from the HW ROT. An L4 could extend trust further into runtime, e.g., IMA/TCG log events that resulted in extending PCRs. Please poke holes on that simplified, imperfect summary.

The gist of this question is not whether the control plane should be trusted or not (in L2+, it's not) but whether the event which the control plane emits to dispatch the build influences the build, _even if the `BuildEnv` is correct_, and I think this scenario is plausible.

So I agree with @paveliak on this front:

>  For example, we have to trust build request message that initiates the build. [...] generally we should include into the provenance all input messages coming from the control plane that can influence build results.

Such an event might contain information that users might not feel comfortable including in provenance. It should be possible for a user to understand how the build platform trusts its control plane. How does it talk to it? Over a host interface, or over the network? Over TLS? Hardcoded endpoints? Can the roots of trust be tampered with? Can it be MITMed? _Is the dispatch event itself signed_? If so, how does the build platform get keys to verify it.

Of course there's an approach where users can have even more control, and they can enforce that the build doesn't start (e.g., git reference is never checked out) unless they can verify the dispatch event is what they expected.

For example, in Azure Confidential Computing, [enforcement policies](https://github.com/Azure/azure-cli-extensions/blob/main/src/confcom/azext_confcom/README.md#security-policy-rules-documentation) are authored client-side, and the control plane passes them through to the underlying compute platform and eventually into the TCB. Inside the TCB, a minimal and well-understood component obtains a hardware-attested measurement of that policy which is non-falsifiable and presented to the user. The user then decides how to proceed, for example by sealing a key to that state as part of a secure key release protocol. Some aspects of this problem remind me of that approach, but it introduces significant complexity compared to the alternative where we can say `BuildEnv` L2+ distrusts the control plane, creating requirements for build platforms as to how they establish trust with their control plane, and possibly recording the dispatch events as part of provenance.

**Comment by arewm** [comment]
Looking at https://slsa.dev/spec/v1.1/terminology#build-model, there is a separation between control plan and build environment. I guess from my perspective, this clear boundary doesn't always exist. 

If we take the build platform that I have the most experience with (Tekton), the line between control plane and build environment is not always clear. If we have a system that enables specific users-defined actions to produce artifacts (i.e. via a Containerfile), where is that separation? Do we, for example, classify a single TaskRun as a build environment or is a process within that task the build environment (for example a buildah process building the image)? Or is the control plane only the Tekton orchestration and all of the triggered workloads the build environment?

Even if Tekton + Chains is the control plane, what verification should we have on those workloads?

**Comment by marcelamelara** [comment]
> Looking at https://slsa.dev/spec/v1.1/terminology#build-model, there is a separation between control plan and build environment. I guess from my perspective, this clear boundary doesn't always exist.

@paveliak and I were just having this exact discussion on Friday. We realized that even for VM-based build platforms like Github actions, it's not always clear what the boundary is because there are control plane components running inside of every build environment. See: https://slsa.dev/spec/draft/terminology#build-environment-model

@arewm I'm not particularly familiar with Tekton Chains, so I definitely appreciate this perspective. Is there a clear boundary for the build image, i.e. the container image used to spawn new containers to run builds?

The way we currently get around this in the BuildEnv track is that we say that the control plane (still a black box) is trusted, so we aren't requiring provenance for it. Because the track _does_ care about integrity of the build environment, any control plane-related components within the build _image_ do need to be integrity measured and included in the image's provenance. Is this sufficient?

**Comment by bureado** [comment]
> Because the track does care about integrity of the build environment, any control plane-related components within the build image used to spawn the environment do need to be integrity measured and included in the image's provenance. Is this sufficient?

Yes, I'd think so. And in more mature levels, adequate separation of those components possibly into a distinct privilege level, and runtime integrity monitoring of them, are additive and increase the ability of the users to make their trust in those control plane components more explicit.
